Client/Frontend/Brave Today/Composer/FeedDataSource.swift

```diff
@@ -0,0 +1,581 @@
+        let filename = resource.filename
+        let fileManager = FileManager.default
+        let deferred = Deferred<Data?>(value: nil, defaultQueue: .main)
+        let cachedPath = fileManager.getOrCreateFolder(name: Self.cacheFolderName)?.appendingPathComponent(filename).path
+        if (loadExpiredData || !isResourceExpired(resource)),
+            let cachedPath = cachedPath,
+            fileManager.fileExists(atPath: cachedPath) {
+            todayQueue.async {
+                if let cachedContents = fileManager.contents(atPath: cachedPath) {
+                    deferred.fill(cachedContents)
+                } else {
+                    deferred.fill(nil)
+                }
+            }
+        } else {
+            deferred.fill(nil)
+        }
+        return deferred
+    }
+    
+    /// Load a Brave Today resource either from a file cache or the web
+    ///
+    /// The `filename` provided will be appended as a path component to the request URL, and be used to
+    /// fetch the cache and save the response so it should include the full path for the endpoint (For
+    /// example: `sources.json`)
+    ///
+    /// Cache lifetime will be based on the modification date of the cached file. Data downloaded from the web
+    /// will only be cached if it is successfully decoded into the given `DataType`.
+    private func loadResource<DataType>(
+        _ resource: TodayResource,
+        decodedTo: DataType.Type
+    ) -> Deferred<Result<DataType, Error>> where DataType: Decodable {
+        let filename = resource.filename
+        return cachedResource(resource)
+            .bind({ [weak self] data -> Deferred<Result<Data, Error>> in
+                guard let self = self else { return .init() }
+                if let data = data {
+                    return .init(value: .success(data), defaultQueue: .main)
+                }
+                let deferred = Deferred<Result<Data, Error>>(value: nil, defaultQueue: .main)
+                self.session.dataRequest(with: Self.baseURL.appendingPathComponent(filename)) { data, response, error in
+                    if let error = error {
+                        deferred.fill(.failure(error))
+                        return
+                    }
+                    guard let data = data else { return }
+                    deferred.fill(.success(data))
+                }
+                return deferred
+            })
+            .mapQueue(todayQueue) { result in
+                switch result {
+                case .success(let data):
+                    do {
+                        let decodedResource = try self.decoder.decode(DataType.self, from: data)
+                        if !data.isEmpty {
+                            if !FileManager.default.writeToDiskInFolder(data, fileName: filename, folderName: Self.cacheFolderName) {
+                                logger.error("Failed to write sources to disk")
+                            }
+                        }
```

Comment:
if `data.isEmpty` -> return .failure()

Reply:
This write is technically for cache purposes… Michal brought it up earlier, I'm not sure if failure to write a cache to disk should warrant a failure